From 59dc7ec6e560faf78267ef36703be081ddc50520 Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Wed, 22 Oct 2025 15:23:27 +0200
Subject: [PATCH] counter: stm32-timer-cnt: add support to count on external
 trigger signal

STM32 timers can count on ETR (external trigger) pin or internal etr
sources. So define ETR signal, and add relevant signal action to
count on it, through the "increase" function.
Rename the counter synapse action used for clock signal, to none/rising
and use it for ETR signal.

Example to enable counting on ETR pin:
$ cd /sys/bus/counter/devices/counter0/count0/
$ cat ../signal5/name
ETR
$ echo rising edge > signal5_action
$ echo 1 > enable

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Change-Id: I56c8813afd0642cd191574cff1e4343b5f0fb8a1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/487189
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/counter/stm32-timer-cnt.c | 33 ++++++++++++++++++++++++++++---
 1 file changed, 30 insertions(+), 3 deletions(-)

--- a/drivers/counter/stm32-timer-cnt.c
+++ b/drivers/counter/stm32-timer-cnt.c
@@ -29,6 +29,7 @@
 #define STM32_CLOCK_SIG		2
 #define STM32_CH3_SIG		3
 #define STM32_CH4_SIG		4
+#define STM32_ETR_SIG		5
 
 struct stm32_timer_regs {
 	u32 cr1;
@@ -479,7 +480,7 @@ static struct counter_comp stm32_count_e
 	COUNTER_COMP_COUNT_U64("num_overflows", stm32_count_nb_ovf_read, stm32_count_nb_ovf_write),
 };
 
-static const enum counter_synapse_action stm32_clock_synapse_actions[] = {
+static const enum counter_synapse_action stm32_synapse_actions_none_rising[] = {
 	COUNTER_SYNAPSE_ACTION_NONE,
 	COUNTER_SYNAPSE_ACTION_RISING_EDGE,
 };
@@ -540,6 +541,18 @@ static int stm32_action_read(struct coun
 				*action = COUNTER_SYNAPSE_ACTION_RISING_EDGE;
 
 			return 0;
+		case STM32_ETR_SIG:
+			/*
+			 * rising edge on etrf clocks the counter in external clock mode
+			 * In other trigger modes (reset, gated, triggered and combined)
+			 * It influences the counter start/stop/reset, either on edges or
+			 * level. For now, report action "none" for these case, as not
+			 * directly used as clock for counting.
+			 */
+			if (sms == TIM_SMCR_SMS_EXTERNAL_CLOCK_MODE_1 && ts == 7)
+				*action = COUNTER_SYNAPSE_ACTION_RISING_EDGE;
+
+			return 0;
 		case STM32_CH1_SIG:
 		case STM32_CH2_SIG:
 		case STM32_CH3_SIG:
@@ -643,6 +656,11 @@ static int stm32_action_write(struct cou
 			sms = TIM_SMCR_SMS_EXTERNAL_CLOCK_MODE_1;
 			ts = 6;
 			break;
+		case STM32_ETR_SIG:
+			/* rising edge on etrf clocks the counter */
+			sms = TIM_SMCR_SMS_EXTERNAL_CLOCK_MODE_1;
+			ts = 7;
+			break;
 		default:
 			dev_err(counter->parent, "Action [%d] not supported for signal [%d]\n",
 				action, signal_id);
@@ -967,6 +985,10 @@ static struct counter_signal stm32_signa
 		.ext = stm32_count_channel_ext,
 		.num_ext = ARRAY_SIZE(stm32_count_channel_ext),
 	},
+	{
+		.id = STM32_ETR_SIG,
+		.name = "ETR",
+	},
 };
 
 static struct counter_synapse stm32_count_synapses[] = {
@@ -981,8 +1003,8 @@ static struct counter_synapse stm32_coun
 		.signal = &stm32_signals[STM32_CH2_SIG]
 	},
 	{
-		.actions_list = stm32_clock_synapse_actions,
-		.num_actions = ARRAY_SIZE(stm32_clock_synapse_actions),
+		.actions_list = stm32_synapse_actions_none_rising,
+		.num_actions = ARRAY_SIZE(stm32_synapse_actions_none_rising),
 		.signal = &stm32_signals[STM32_CLOCK_SIG]
 	},
 	{
@@ -995,6 +1017,11 @@ static struct counter_synapse stm32_coun
 		.num_actions = ARRAY_SIZE(stm32_channel_synapse_actions),
 		.signal = &stm32_signals[STM32_CH4_SIG]
 	},
+	{
+		.actions_list = stm32_synapse_actions_none_rising,
+		.num_actions = ARRAY_SIZE(stm32_synapse_actions_none_rising),
+		.signal = &stm32_signals[STM32_ETR_SIG]
+	},
 };
 
 static struct counter_count stm32_counts = {
